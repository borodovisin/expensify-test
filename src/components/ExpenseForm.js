import React, { Component } from 'react';

export default class ExpenseForm extends Component {
    constructor(props){ 
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            error: ''
        }
    }

    onDescriptionChanges = e => {
        const description =  e.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChanges = e => {
        const note =  e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChanges = e => {
        const amount = e.target.value;
        if (!amount ||  amount.match(/^\d*(\.\d{0,2})?$/)) this.setState(() => ({amount}));
    }

    onSubmit = e => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "Description and amount are mandatory fields" }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder="Description" autoFocus value={this.state.description}
                        onChange={this.onDescriptionChanges} />
                    <input type="text" placeholder="Amount" value={this.state.amount} 
                    onChange={this.onAmountChanges}/>
                    <textarea placeholder="Add a note for your expense (optional)" onChange={this.onNoteChanges} value={this.state.note} />
                    <button>Add Expense</button>
                    {this.state.error && <p>{this.state.error}</p>}
                </form>
            </div>
        );
    }
}