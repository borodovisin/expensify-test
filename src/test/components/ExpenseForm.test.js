import React from 'react';
import { shallow } from 'enzyme';
import ExpenseForm from '../../components/ExpenseForm';

test('should render ExpenseForm correctly', () => {
    const wrapper = shallow(<ExpenseForm />);
    expect(wrapper).toMatchSnapshot();
})

test('should render error for invalid form submission', () => {
    const wrapper = shallow(<ExpenseForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
})

test('should set description on input change', () => {
    const value = 'New Description';
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: { value: value}
    })
    expect(wrapper.state('description')).toBe(value)
})

test('should not set amount if invalid input', () => {
    const value = '12.122';
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: { value: value}
    })
    expect(wrapper.state('description')).toBe('12.122');
})

test('should set note on textarea change', () => {
    const value = 'New note value';
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('textarea').simulate('change', {
        target: { value: value}
    })
    expect(wrapper.state('note')).toBe(value)
})

test('should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<ExpenseForm expense={{amount: 195, description: "Gum", id: 1, note: ""}} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    })
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({amount: 195, description: "Gum", note: ""});
})

