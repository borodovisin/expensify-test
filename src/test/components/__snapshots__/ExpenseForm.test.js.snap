// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <button>
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <button>
      Add Expense
    </button>
    <p>
      Description and amount are mandatory fields
    </p>
  </form>
</div>
`;
